pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,683 @@
+				continue
+			}
+
+			// filter or load manifest list as appropriate
+			srcManifests, srcManifest, srcDigest, err := processManifestList(ctx, srcDigest, srcManifest, manifests, src.ref, o.includeDescriptor)
+			if err != nil {
+				digestErrs = append(digestErrs, retrieverError{src: src.ref, err: err})
+				continue
+			}
+			if len(srcManifests) == 0 {
+				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
+				continue
+			}
+
+			for _, dst := range pushTargets {
+				// if we are going to be using cross repository mount, get a token that covers the src
+				if src.ref.Registry == dst.ref.Registry {
+					toContext = toContext.WithScopes(auth.RepositoryScope{Repository: src.ref.RepositoryName(), Actions: []string{"pull"}})
+				}
+
+				toRepo, err := o.Repository(ctx, toContext, creds, dst.t, dst.ref)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to connect to %s: %v", dst.ref, err)})
+					continue
+				}
+
+				canonicalTo := toRepo.Named()
+				toManifests, err := toRepo.Manifests(ctx)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to access destination image %s manifests: %v", src.ref, err)})
+					continue
+				}
+
+				var mustCopyLayers bool
+				switch {
+				case o.Force:
+					mustCopyLayers = true
+				case src.ref.Registry == dst.ref.Registry && canonicalFrom.String() == canonicalTo.String():
+					// if the source and destination repos are the same, we don't need to copy layers unless forced
+				default:
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
+						case *manifestlist.DeserializedManifestList:
+							// we do not need to upload layers in a manifestlist
+							continue
+						default:
+							digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("the manifest type %T is not supported", srcManifest)})
+							continue
+						}
+
+						for _, blob := range srcManifest.References() {
+							blobSource, err := reference.WithDigest(canonicalFrom, blob.Digest)
+							if err != nil {
+								digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unexpected error building named digest: %v", err)})
+								continue
+							}
+
+							// if we aren't forcing upload, skip the blob copy
+							if !o.Force {
+								_, err := toBlobs.Stat(ctx, blob.Digest)
```

Comment:
To save a few queries, a set of blobs uploaded to particular repo could be created. I'd expect some shared blobs among different manifests of a single manifest list.

Reply:
Yes.  I think when we parallelize this we'll want to do it.